File: apps/desktop/src/lib/api/chats.ts
import { locale } from "../../i18n";
/** Conversations, their settings, and the streaming send. */

import { getApiConfig } from "../config";
import { request } from "./client";
import type { ChatDetail, ChatMessage, ChatSearchResult, ChatSummary, ReplySettings, Resolution, SummarizeResult, TaskStatus, ToolCall } from "../types";

/** Chats whose messages or titles hold every word of `q`; Arabic is folded on the agent. */
export async function searchChats(q: string, workspaceId?: string | null, signal?: AbortSignal): Promise<ChatSearchResult[]> {
  const params = new URLSearchParams({ q });
  if (workspaceId) params.set("workspace_id", workspaceId);
  return request<ChatSearchResult[]>(`/chats/search?${params}`, { signal });
}

export async function listChats(workspaceId?: string | null): Promise<ChatSummary[]> {
  const query = workspaceId ? `?workspace_id=${encodeURIComponent(workspaceId)}` : "";
  return request<ChatSummary[]>(`/chats${query}`);
}

export async function createChat(modelId?: string, workingDir?: string | null, workspaceId?: string | null): Promise<ChatDetail> {
  return request<ChatDetail>("/chats", {
    method: "POST",
    body: JSON.stringify({ model_id: modelId, working_dir: workingDir || undefined, workspace_id: workspaceId ?? null }),
  });
}

/** The whole chat as one file to share: self-contained HTML (both themes) or Markdown. */
export async function exportChat(id: string, format: "html" | "md"): Promise<string> {
  const { baseUrl, token } = await getApiConfig();
  const res = await fetch(`${baseUrl}/chats/${id}/export?format=${format}`, {
    headers: { Authorization: `Bearer ${token}`, "Accept-Language": locale() },
  });
  if (!res.ok) throw new Error(await res.text());
  return res.text();
}

export async function setChatFolder(id: string, workingDir: string | null): Promise<ChatSummary> {
  return request<ChatSummary>(`/chats/${id}`, { method: "PATCH", body: JSON.stringify({ working_dir: workingDir ?? "" }) });
}

export async function resolveChatPermission(chatId: string, requestId: string, resolution: "approved" | "denied"): Promise<void> {
  await request<void>(`/chats/${chatId}/permissions/${requestId}?resolution=${resolution}`, { method: "POST" });
}

export async function getChat(id: string): Promise<ChatDetail> {
  return request<ChatDetail>(`/chats/${id}`);
}

export async function renameChat(id: string, title: string): Promise<ChatSummary> {
  return request<ChatSummary>(`/chats/${id}`, { method: "PATCH", body: JSON.stringify({ title }) });
}

export async function setChatPinned(id: string, pinned: boolean): Promise<ChatSummary> {
  return request<ChatSummary>(`/chats/${id}`, { method: "PATCH", body: JSON.stringify({ pinned }) });
}

export async function updateChatSettings(id: string, settings: ReplySettings): Promise<ChatSummary> {
  return request<ChatSummary>(`/chats/${id}`, { method: "PATCH", body: JSON.stringify({ settings }) });
}

/** Folds the older turns into a summary so later messages stop resending the whole history. */
export async function summarizeChat(id: string, modelId: string, keep = 4): Promise<SummarizeResult> {
  return request<SummarizeResult>(`/chats/${id}/summarize`, {
    method: "POST",
    body: JSON.stringify({ model_id: modelId, keep }),
  });
}

export async function deleteChatMessage(chatId: string, messageId: string): Promise<void> {
  await request<void>(`/chats/${chatId}/messages/${messageId}`, { method: "DELETE" });
}

/** Drops a message and everything after it — how editing a question starts. */
export async function truncateChatFrom(chatId: string, messageId: string): Promise<ChatDetail> {
  return request<ChatDetail>(`/chats/${chatId}/messages/${messageId}/truncate`, { method: "POST" });
}

/** Copies the chat up to a message into a new one (leaving this one untouched). */
export async function forkChat(chatId: string, untilMessageId?: string): Promise<ChatDetail> {
  return request<ChatDetail>(`/chats/${chatId}/fork`, {
    method: "POST",
    body: JSON.stringify({ until_message_id: untilMessageId ?? null }),
  });
}

export async function deleteChat(id: string): Promise<void> {
  await request<void>(`/chats/${id}`, { method: "DELETE" });
}

export type ChatStreamEvent =
  | { type: "start"; user_message: ChatMessage; title: string }
  | { type: "reasoning"; text: string }
  | { type: "delta"; text: string }
  | { type: "tool_call"; id: string; tool: string; args: Record<string, unknown> }
  | { type: "tool_result"; id: string; ok: boolean; output: string }
  | { type: "permission"; id: string; call: ToolCall }
  | { type: "permission_resolved"; id: string; resolution: Resolution }
  | { type: "task_created"; task: { id: string; title: string; status: TaskStatus } }
  | { type: "usage"; prompt_tokens: number; completion_tokens: number; cached_tokens: number; cost_usd: number }
  | { type: "done"; message: ChatMessage }
  | { type: "stopped" }
  | { type: "error"; message: string };

/**
 * Sends a message and streams the reply (Server-Sent Events over fetch). Aborting `signal`
 * only stops *watching*: the reply keeps being written on the agent, and
 * `attachChatStream` picks it up again. `stopChat` is what stops it.
 */
export async function sendChatMessage(
  chatId: string,
  content: string,
  modelId: string,
  attachmentIds: string[],
  onEvent: (event: ChatStreamEvent) => void,
  signal?: AbortSignal,
): Promise<void> {
  const { baseUrl, token } = await getApiConfig();
  const res = await fetch(`${baseUrl}/chats/${chatId}/messages`, {
    method: "POST",
    cache: "no-store",
    signal,
    headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json", "Accept-Language": locale() },
    body: JSON.stringify({ content, model_id: modelId, attachment_ids: attachmentIds }),
  });
  await readEvents(res, onEvent);
}

/**
 * Rejoins the reply being written in a chat, replaying it from its first event.
 * Resolves `false` when the chat has no reply in progress.
 */
export async function attachChatStream(
  chatId: string,
  onEvent: (event: ChatStreamEvent) => void,
  signal?: AbortSignal,
): Promise<boolean> {
  const { baseUrl, token } = await getApiConfig();
  const res = await fetch(`${baseUrl}/chats/${chatId}/stream`, {
    cache: "no-store",
    signal,
    headers: { Authorization: `Bearer ${token}`, "Accept-Language": locale() },
  });
  if (res.status === 204) return false;
  await readEvents(res, onEvent);
  return true;
}

/** Stops the reply being written; what it wrote so far is kept. */
export async function stopChat(chatId: string): Promise<boolean> {
  const out = await request<{ stopped: boolean }>(`/chats/${chatId}/stop`, { method: "POST" });
  return out.stopped;
}

async function readEvents(res: Response, onEvent: (event: ChatStreamEvent) => void): Promise<void> {
  if (!res.ok || !res.body) {
    const detail = await res.json().catch(() => null);
    throw new Error(detail?.detail ?? `${res.status} ${res.statusText}`);
  }

  const reader = res.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  for (;;) {
    const { value, done } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    let boundary = buffer.indexOf("\n\n");
    while (boundary !== -1) {
      const frame = buffer.slice(0, boundary);
      buffer = buffer.slice(boundary + 2);
      const data = frame
        .split("\n")
        .filter((line) => line.startsWith("data:"))
        .map((line) => line.slice(5).trimStart())
        .join("\n");
      if (data) onEvent(JSON.parse(data) as ChatStreamEvent);
      boundary = buffer.indexOf("\n\n");
    }
  }
}

File: apps/agent/rafiq_agent/api/chats.py
"""Chat endpoints.

Routes only: validate input, call `core.chat_service`, shape the response. The turn itself
(context building, the agent loop, the SSE events) lives in the service.
"""

from datetime import UTC, datetime
from urllib.parse import quote

from fastapi import APIRouter, Depends, HTTPException, Query, Response
from fastapi.responses import StreamingResponse
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from rafiq_agent.api.deps import require_token
from rafiq_agent.core.chat_service import (
    ChatError,
    ChatTurn,
    active_turn,
    fallback_of,
    provider_for,
    resolve_permission,
    settings_of,
    stop_turn,
    summarize,
)
from rafiq_agent.core.agent_runtime import load_settings
from rafiq_agent.core.export_html import render as render_html
from rafiq_agent.core.prompts import DEFAULT_TITLE
from rafiq_agent.core.tasks_service import TaskCreateError, resolve_working_dir
from rafiq_agent.i18n import tr
from rafiq_agent.schemas.chats import (
    ChatCreate,
    ChatDetailOut,
    ChatFork,
    ChatSummaryOut,
    ChatUpdate,
    MessageCreate,
    SummarizeIn,
    SummarizeOut,
)
from rafiq_agent.storage.db import SessionLocal, get_session
from rafiq_agent.storage.models import Chat, ChatMessage, LlmModel

router = APIRouter(prefix="/chats", tags=["chats"], dependencies=[Depends(require_token)])


def _now() -> datetime:
    return datetime.now(UTC)


def _checked_dir(raw: str | None) -> str | None:
    if not raw:
        return None
    try:
        directory = resolve_working_dir(raw)
    except TaskCreateError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return str(directory) if directory else None


def _http(error: ChatError) -> HTTPException:
    return HTTPException(status_code=error.status, detail=str(error))


@router.get("", response_model=list[ChatSummaryOut])
async def list_chats(
    workspace_id: str | None = Query(None), session: AsyncSession = Depends(get_session)
) -> list[ChatSummaryOut]:
    counts = (
        select(ChatMessage.chat_id, func.count(ChatMessage.id).label("n"))
        .group_by(ChatMessage.chat_id)
        .subquery()
    )
    rows = await session.execute(
        select(Chat, func.coalesce(counts.c.n, 0))
        .outerjoin(counts, counts.c.chat_id == Chat.id)
        # Design sessions live on the designs page; they'd only be noise here.
        .where(func.coalesce(Chat.mode, "chat") != "design")
        .where(Chat.workspace_id == workspace_id if workspace_id else True)
        .order_by(Chat.pinned.desc(), Chat.updated_at.desc())
    )
    out = []
    for chat, count in rows.all():
        summary = ChatSummaryOut.model_validate(chat)
        summary.message_count = count
        summary.streaming = active_turn(chat.id) is not None
        out.append(summary)
    return out


@router.post("", response_model=ChatDetailOut, status_code=201)
async def create_chat(body: ChatCreate, session: AsyncSession = Depends(get_session)) -> ChatDetailOut:
    chat = Chat(
        title=tr(DEFAULT_TITLE),
        model_id=body.model_id,
        working_dir=_checked_dir(body.working_dir),
        workspace_id=body.workspace_id or None,
        # Only the one field: everything else stays on its schema default, so a default we
        # improve later still reaches chats created today.
        settings={"saver": (await load_settings()).token_saver},
    )
    session.add(chat)
    await session.commit()
    await session.refresh(chat)
    return ChatDetailOut(**ChatSummaryOut.model_validate(chat).model_dump(), messages=[])


@router.get("/{chat_id}", response_model=ChatDetailOut)
async def get_chat(chat_id: str, session: AsyncSession = Depends(get_session)) -> ChatDetailOut:
    result = await session.execute(
        select(Chat).where(Chat.id == chat_id).options(selectinload(Chat.messages))
    )
    chat = result.scalar_one_or_none()
    if not chat:
        raise HTTPException(status_code=404, detail="chat not found")
    out = ChatDetailOut.model_validate(chat)
    out.streaming = active_turn(chat_id) is not None
    return out


@router.get("/{chat_id}/export")
async def export_chat(
    chat_id: str, format: str = Query("html"), session: AsyncSession = Depends(get_session)
) -> Response:
    """The whole chat as one file to share: HTML (self-contained, both themes) or Markdown."""
    result = await session.execute(
        select(Chat).where(Chat.id == chat_id).options(selectinload(Chat.messages))
    )
    chat = result.scalar_one_or_none()
    if not chat:
        raise HTTPException(status_code=404, detail="chat not found")
    models = {m.id: m.name for m in (await session.execute(select(LlmModel))).scalars().all()}
    messages = [
        {
            "role": m.role,
            "content": m.content,
            "parts": m.parts,
            "created_at": m.created_at,
            "model_id": m.model_id,
        }
        for m in chat.messages
    ]
    # Headers are latin-1: an Arabic title goes in the RFC 5987 form, with an ASCII fallback.
    safe = "".join(ch if ch.isalnum() or ch in " -_" else "_" for ch in chat.title)[:60].strip() or "chat"
    ascii_name = safe.encode("ascii", "ignore").decode().strip() or "chat"
    encoded = quote(safe)
    if format == "md":
        lines = [f"# {chat.title}", ""]
        for m in messages:
            who = f"**{tr('أنا')}**" if m["role"] == "user" else f"**{models.get(m['model_id'] or '', tr('رفيق'))}**"
            lines += [who, "", m["content"] or "", ""]
        body = "\n".join(lines)
        return Response(
            body,
            media_type="text/markdown; charset=utf-8",
            headers={"Content-Disposition": f"attachment; filename=\"{ascii_name}.md\"; filename*=UTF-8''{encoded}.md"},
        )
    html = render_html(chat.title, messages, models)
    return Response(
        html,
        media_type="text/html; charset=utf-8",
        headers={"Content-Disposition": f"attachment; filename=\"{ascii_name}.html\"; filename*=UTF-8''{encoded}.html"},
    )


@router.patch("/{chat_id}", response_model=ChatSummaryOut)
async def update_chat(
    chat_id: str, body: ChatUpdate, session: AsyncSession = Depends(get_session)
) -> ChatSummaryOut:
    chat = await session.get(Chat, chat_id)
    if not chat:
        raise HTTPException(status_code=404, detail="chat not found")
    if body.title is not None and body.title.strip():
        chat.title = body.title.strip()[:120]
    if body.model_id is not None:
        chat.model_id = body.model_id
    if body.pinned is not None:
        chat.pinned = body.pinned
    if body.working_dir is not None:
        chat.working_dir = _checked_dir(body.working_dir)
    if body.settings is not None:
        chat.settings = body.settings.model_dump()
    await session.commit()
    await session.refresh(chat)
    return ChatSummaryOut.model_validate(chat)


@router.delete("/{chat_id}", status_code=204)
async def delete_chat(chat_id: str, session: AsyncSession = Depends(get_session)) -> None:
    result = await session.execute(
        select(Chat).where(Chat.id == chat_id).options(selectinload(Chat.messages))
    )
    chat = result.scalar_one_or_none()
    if not chat:
        raise HTTPException(status_code=404, detail="chat not found")
    stop_turn(chat_id)
    await session.delete(chat)
    await session.commit()


@router.post("/{chat_id}/permissions/{request_id}", status_code=202)
async def resolve_chat_permission(chat_id: str, request_id: str, resolution: str) -> dict[str, bool]:
    if resolution not in ("approved", "denied"):
        raise HTTPException(status_code=400, detail="resolution must be 'approved' or 'denied'")
    return {"resolved": resolve_permission(request_id, resolution)}


@router.post("/{chat_id}/summarize", response_model=SummarizeOut)
async def summarize_chat(chat_id: str, body: SummarizeIn) -> SummarizeOut:
    """Condenses the chat so later turns send a short summary instead of the whole history."""
    async with SessionLocal() as session:
        result = await session.execute(
            select(Chat).where(Chat.id == chat_id).options(selectinload(Chat.messages))
        )
        chat = result.scalar_one_or_none()
        if not chat:
            raise HTTPException(status_code=404, detail="chat not found")
        model = await session.get(LlmModel, body.model_id or chat.model_id or "")
        if not model:
            raise HTTPException(status_code=400, detail=tr("اختار نموذج أول عشان ألخّص فيه."))

        llm = provider_for(model, settings_of(chat), await fallback_of(session, model))
        try:
            report = await summarize(chat, list(chat.messages), llm, max(0, body.keep))
        except ChatError as exc:
            raise _http(exc) from exc
        except Exception as exc:  # noqa: BLE001 - provider failures become a readable message
            from rafiq_agent.llm.discovery import friendly_error

            raise HTTPException(status_code=502, detail=friendly_error(exc)) from exc

        chat.updated_at = _now()
        await session.commit()
        return SummarizeOut(summary_until=chat.summary_until or "", **report)


@router.delete("/{chat_id}/messages/{message_id}", status_code=204)
async def delete_message(chat_id: str, message_id: str, session: AsyncSession = Depends(get_session)) -> None:
    """Used by /أعد — drops the last reply before the same prompt is sent again."""
    message = await session.get(ChatMessage, message_id)
    if not message or message.chat_id != chat_id:
        raise HTTPException(status_code=404, detail="message not found")
    await session.delete(message)
    await session.commit()


@router.post("/{chat_id}/messages/{message_id}/truncate", response_model=ChatDetailOut)
async def truncate_from(
    chat_id: str, message_id: str, session: AsyncSession = Depends(get_session)
) -> ChatDetailOut:
    """Drops this message and everything after it — how editing a question works: the
    chat goes back to just before it, then the new wording is sent."""
    message = await session.get(ChatMessage, message_id)
    if not message or message.chat_id != chat_id:
        raise HTTPException(status_code=404, detail="message not found")
    stop_turn(chat_id)
    result = await session.execute(
        select(Chat).where(Chat.id == chat_id).options(selectinload(Chat.messages))
    )
    chat = result.scalar_one_or_none()
    if not chat:
        raise HTTPException(status_code=404, detail="chat not found")
    cutoff = message.created_at
    for m in list(chat.messages):
        if m.created_at > cutoff or m.id == message_id:
            await session.delete(m)
    await session.commit()
    await session.refresh(chat)
    return await get_chat(chat_id, session)


@router.post("/{chat_id}/fork", response_model=ChatDetailOut, status_code=201)
async def fork_chat(
    chat_id: str, body: ChatFork, session: AsyncSession = Depends(get_session)
) -> ChatDetailOut:
    """Copies the chat up to a message into a new one, so a different direction can be
    tried without losing this one."""
    result = await session.execute(
        select(Chat).where(Chat.id == chat_id).options(selectinload(Chat.messages))
    )
    chat = result.scalar_one_or_none()
    if not chat:
        raise HTTPException(status_code=404, detail="chat not found")
    history = sorted(chat.messages, key=lambda m: m.created_at)
    if body.until_message_id:
        index = next((i for i, m in enumerate(history) if m.id == body.until_message_id), None)
        if index is None:
            raise HTTPException(status_code=404, detail="message not found")
        history = history[: index + 1]

    fork = Chat(
        title=tr("نسخة: {0}", chat.title)[:120],
        model_id=chat.model_id,
        working_dir=chat.working_dir,
        settings=chat.settings,
        mode=chat.mode,
        # The summary only describes turns that were kept whole; a partial copy re-summarises.
        summary=chat.summary if len(history) == len(chat.messages) else None,
        summary_until=chat.summary_until if len(history) == len(chat.messages) else None,
    )
    session.add(fork)
    await session.flush()
    for m in history:
        session.add(
            ChatMessage(
                chat_id=fork.id,
                role=m.role,
                content=m.content,
                reasoning=m.reasoning,
                model_id=m.model_id,
                parts=m.parts,
                attachments=m.attachments,
                created_at=m.created_at,
            )
        )
    await session.commit()
    return await get_chat(fork.id, session)


def _sse(turn: ChatTurn) -> StreamingResponse:
    return StreamingResponse(
        turn.subscribe(), media_type="text/event-stream", headers={"X-Accel-Buffering": "no"}
    )


@router.post("/{chat_id}/messages")
async def send_message(chat_id: str, body: MessageCreate) -> StreamingResponse:
    """Starts a reply and streams it. The reply keeps going if the stream is closed."""
    turn = ChatTurn(chat_id, body.content, body.model_id, body.attachment_ids)
    try:
        await turn.prepare()
        await turn.start()
    except ChatError as exc:
        raise _http(exc) from exc
    return _sse(turn)


@router.get("/{chat_id}/stream", response_model=None)
async def reattach(chat_id: str) -> StreamingResponse | Response:
    """The reply being written in this chat, from its first event; 204 when there is none."""
    turn = active_turn(chat_id)
    if turn is None:
        return Response(status_code=204)
    return _sse(turn)


@router.post("/{chat_id}/stop", status_code=202)
async def stop_reply(chat_id: str) -> dict[str, bool]:
    """Stops the reply being written; what it produced so far is kept."""
    return {"stopped": stop_turn(chat_id)}
